### PR TITLE
Add e2e test workflow

### LINKED ISSUE #8: Add e2e tests to be run on actual Azure CVM
### Static setup using terraform (or whatever is convenient)
- Create a RG that will run the CVM tests.
- Permissions to bring up a CVM in a RG only will be given via [federated auth](https://docs.github.com/en/actions/deployment/security-hardening-your-deployments/configuring-openid-connect-in-azure).

## For each CI run

- Clone code and build binaries locally in the normal runners.
- Bring up a CVM, install dependencies for dynamically linked libraries.
- Copy (scp) the binaries over the the CVM.
- Run the example code and other tests. (replace the code that uses local check-in artifacts with the artifacts available on the machine)
- Clean up the CVM.

### PR DESCRIPTION
**Pull request by mkulke**
# Add e2e test workflow

fixes #8 

This adds an e2e test as a workflow that will
- Build the example project
- Spawn a CVM
- Copy the example bin
- Execute the example bin
- Delete the CVM + aux resources

It will use azure via federated credentials. The AD app needs 2 role assignments: `Network Contributor` + `Virtual Machine Contributor`. The app is scoped to the resource group `azure-cvm-tooling-ci`.

The arm template + makefile was modified to allow implicit creation (and cleanup) of a vnet + subnet, if no subnetId is given.

Other than specified in the issue, it will not create a temporary resource group, since we want to scope the AD app to a given resource group.

The job is executed on the `restricted` environment, which means that it requires repo maintainers to approve a workflow run.

## How to use

Open a PR or push to main

## Testing done

Completed a [successful run](https://github.com/kinvolk/azure-cvm-tooling/actions/runs/4805725330/jobs/8552426366).

### CODE REVIEW

File: .github/workflows/e2e.yaml [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+  contents: read
+
+jobs:
+  build-and-run-example:
+    runs-on: ubuntu-latest
+    environment: restricted
+    steps:
+    - name: Create resource suffix
+      run: >
+        echo "SUFFIX=$(echo $RANDOM | md5sum | head -c6)"
```
Comment (surajssd) [inline .github/workflows/e2e.yaml:20]: Is there a way to make this `SUFFIX` look like `<pr number>-<random number>`, just for convenience. If the cleanup fails and we had to do manual clean up.

File: az-snp-vtpm/Makefile [hunk 2]
```diff
@@ -39,4 +35,7 @@ delete:
 		--yes && \
 	az network public-ip delete \
 		--resource-group $(CVM_RESOURCE_GROUP) \
-		--name $(VM_NAME)-ip
+		--name $(VM_NAME)-ip && \
```
Comment (surajssd) [inline az-snp-vtpm/Makefile:38]: Can we make deletion idempotent and resilient? Using `&&` makes it prone frequent failures.

File: .github/workflows/e2e.yaml [hunk 3]
```diff
@@ -0,0 +1,89 @@
+name: e2e
+
+on:
+  push:
+    branches: [ "main" ]
```
Comment (surajssd) [inline .github/workflows/e2e.yaml:5]: For merge to main this should automatically run isn't it?

**Review by surajssd — COMMENTED** [review]
Re: Static Infrastructure: Is it just permissions and a RG?

File: az-snp-vtpm/Makefile (hunk 2 above)
Comment (mkulke) [reply az-snp-vtpm/Makefile:38]: I think the az cli works like this ootb, if you delete a non-existing resource exit code will be still 0

File: .github/workflows/e2e.yaml (hunk 1 above)
Comment (mkulke) [reply .github/workflows/e2e.yaml:20]: > Re: Static Infrastructure: Is it just permissions and a RG?

yes, also the `restricted` env needs client_id, tenant_id, subscription_id set as secrets and on the azure side there is an app + federated credentials + role assignments.

File: az-snp-vtpm/Makefile (hunk 2 above)
Comment (surajssd) [reply az-snp-vtpm/Makefile:38]: But for the failed command the chain will stop isn't it? Leaving dangling resources?

Comment (mkulke) [reply az-snp-vtpm/Makefile:38]: Which command do you expect to fail? I think none of the delete commands should fail (unless maybe we can't talk to the api or sth), even if the resources do not exists.

regarding the deletion of the arm template. afaik that's not possible, it's unlike terraform or cloudformation as there is no state beyond deployment.

Comment (surajssd) [reply az-snp-vtpm/Makefile:38]: Alright, I guess let's revisit if we encounter these problems later.

**Pull request merged**